File: backend/app/schemas/trace.py
"""Pydantic schemas for trace request / response validation."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# ── Request ──────────────────────────────────────────────────────────

class TraceCreateRequest(BaseModel):
    """Body of POST /api/traces."""
    code: str = Field(..., min_length=1, description="The C++ source code to trace.")
    problem_type: Literal["grid", "array"] = Field(..., alias="problemType")
    title: str | None = Field(None, max_length=255)

    model_config = {"populate_by_name": True}


# ── Trace step sub-schemas (mirrors the JSON contract) ──────────────

class CellRef(BaseModel):
    row: int | None = None
    col: int | None = None
    index: int | None = None


class CrashInfo(BaseModel):
    type: str
    message: str


class TraceStep(BaseModel):
    step_number: int = Field(..., alias="stepNumber")
    line_of_code: str = Field(..., alias="lineOfCode")
    line_number: int = Field(..., alias="lineNumber")
    action: Literal["read", "write", "compare", "call", "return"]
    cells: list[CellRef] = []
    cell_values: dict[str, list[Any]] = Field(default_factory=dict, alias="cellValues")
    data_structure_state: Any = Field(None, alias="dataStructureState")
    variables: dict[str, Any] = {}
    explanation: str = ""
    call_stack: list[str] = Field(default_factory=list, alias="callStack")
    status: Literal["normal", "crash"] = "normal"
    crash_info: CrashInfo | None = Field(None, alias="crashInfo")

    model_config = {"populate_by_name": True}


# ── Responses ────────────────────────────────────────────────────────

class TraceResponse(BaseModel):
    """Full trace returned to the frontend."""
    id: uuid.UUID
    title: str | None
    problem_type: str = Field(..., alias="problemType")
    initial_state: Any = Field(..., alias="initialState")
    steps: list[TraceStep]
    has_crash: bool = Field(..., alias="hasCrash")
    created_at: datetime = Field(..., alias="createdAt")

    model_config = {"populate_by_name": True, "from_attributes": True}


class TraceListItem(BaseModel):
    """Lightweight trace summary for the history list."""
    id: uuid.UUID
    title: str | None
    problem_type: str = Field(..., alias="problemType")
    has_crash: bool = Field(..., alias="hasCrash")
    created_at: datetime = Field(..., alias="createdAt")

    model_config = {"populate_by_name": True, "from_attributes": True}

File: frontend/src/types/trace.ts
/**
 * TypeScript interfaces matching the JSON trace contract from the PRD.
 */

export interface CellRef {
  row?: number;
  col?: number;
  index?: number;
}

export interface CrashInfo {
  type: string;
  message: string;
}

export interface TraceStep {
  stepNumber: number;
  lineOfCode: string;
  lineNumber: number;
  action: 'read' | 'write' | 'compare' | 'call' | 'return';
  cells: CellRef[];
  cellValues: {
    before: unknown[];
    after: unknown[];
  };
  dataStructureState: string[][] | unknown[];
  variables: Record<string, unknown>;
  explanation: string;
  callStack: string[];
  status: 'normal' | 'crash';
  crashInfo?: CrashInfo;
}

export interface GridState {
  rows: number;
  cols: number;
  data: string[][];
}

export interface ArrayState {
  length: number;
  data: unknown[];
}

export type InitialState = GridState | ArrayState;

export interface TraceResponse {
  id: string;
  title: string | null;
  problemType: 'grid' | 'array';
  initialState: InitialState;
  steps: TraceStep[];
  hasCrash: boolean;
  createdAt: string;
}

export interface TraceListItem {
  id: string;
  title: string | null;
  problemType: 'grid' | 'array';
  hasCrash: boolean;
  createdAt: string;
}

export interface TraceCreateRequest {
  code: string;
  problemType: 'grid' | 'array';
  title?: string;
}
